fuel/utils.py

```diff
@@ -14,6 +16,390 @@
+                               dtype=indexable.dtype)
+            data[indices] = indexable[numpy.array(request)[indices], ...]
+        else:
+            data = indexable[request]
+        return data
+
+    @staticmethod
+    def slice_to_numerical_args(slice_, num_examples):
+        """Translate a slice's attributes into numerical attributes.
+
+        Parameters
+        ----------
+        slice_ : :class:`slice`
+            Slice for which numerical attributes are wanted.
+        num_examples : int
+            Number of examples in the indexable that is to be sliced
+            through. This determines the numerical value for the `stop`
+            attribute in case it's `None`.
+
+        """
+        start = slice_.start if slice_.start is not None else 0
+        stop = slice_.stop if slice_.stop is not None else num_examples
+        step = slice_.step if slice_.step is not None else 1
+        return start, stop, step
+
+    def get_list_representation(self):
+        """Returns this subset's representation as a list of indices."""
+        if self.is_list:
+            return self.list_or_slice
+        else:
+            return self[list(range(self.num_examples))]
+
+    def index_within_subset(self, indexable, subset_request,
+                            safe_hdf5_indexing=True):
+        """Index an indexable object within the context of this subset.
+
+        Parameters
+        ----------
+        indexable : indexable object
+            The object to index through.
+        subset_request : :class:`list` or :class:`slice`
+            List of positive integer indices or slice that constitutes
+            the request *within the context of this subset*. This
+            request will be translated to a request on the indexable
+            object.
+        safe_hdf5_indexing : bool, optional
+            If the indexable is an HDF5 dataset and the request is a list
+            of indices, work around the fancy indexing limitation that
+            requires lists of indices to be sorted by indexing in sorted
+            order and reshuffling the result in the original order.
+            Defaults to `True`.
+
+        """
+        # Translate the request within the context of this subset to a
+        # request to the indexable object
+        if isinstance(subset_request, int):
+            request, = self[[subset_request]]
+        else:
+            request = self[subset_request]
+        # Integer or slice requests can be processed directly.
+        if isinstance(request, int) or hasattr(request, 'step'):
+            return indexable[request]
+        # If the indexable is an HDF5 dataset, it only supports fancy
+        # indexing with sorted lists. As a workaround, if
+        # `safe_hdf5_indexing` is set to `True`, Subset will do the
+        # indexing in sorted order, and reshuffle the result in the
+        # original order.
+        if isinstance(indexable, h5py.Dataset) and safe_hdf5_indexing:
```

Comment:
I am not sure that we should put that much awareness of `H5Py` related issues into `Subset`. What if instead of `safe_hdf5_indexing` we just call the option `safe_indexing` and sort the indices of the subset regardless of what the indexable is? This would make things significantly simpler.